ocaml/xapi/export.ml

```diff
@@ -556,17 +556,26 @@ let metadata_handler (req: Request.t) s _ =
        let task_id = Ref.string_of (Context.get_task_id __context) in
        let read_fd, write_fd = Unix.pipe () in
        let export_error = ref None in
-       let writer_thread = Thread.create (fun () ->
-           (* lock all the VMs before exporting their metadata *)
-           List.iter (fun vm -> lock_vm ~__context ~vm ~task_id `metadata_export) vm_refs;
-           try
-             finally
-               (fun () -> export_metadata ~with_snapshot_metadata:export_snapshots ~preserve_power_state:true ~include_vhd_parents ~__context ~vms:vm_refs write_fd)
-               (fun () ->
-                  Unix.close write_fd;
-                  List.iter (fun vm -> unlock_vm ~__context ~vm ~task_id) vm_refs)
-           with e ->
-             export_error := Some e)
+       let writer_thread = Thread.create (
+           Debug.with_thread_named "metadata export writer thread" (fun () ->
+             (* lock all the VMs before exporting their metadata *)
+             let needs_close = ref true in
+             try
+               List.iter (fun vm -> lock_vm ~__context ~vm ~task_id `metadata_export) vm_refs;
+               finally
+                 (fun () -> export_metadata ~with_snapshot_metadata:export_snapshots ~preserve_power_state:true ~include_vhd_parents ~__context ~vms:vm_refs write_fd)
```

Comment:
Is usually `export_metadata` closing the fd but is not in the case of an exception? That looks like a problem in `export_metadata`. If not, could the closing of the fd not be handled by an outer `finally` without a `needs_close` logic?

Reply:
We are closing it in both situations, but we want to close it sooner in case lock_vm succeeded (i.e. before unlock_vm). This is all very convoluted, and there is still a bug here when lock_vm fails halfway through: we do not unlock any of the VMs that were locked.

So instead I got rid of the 2 close situations and just close it in the finally.
`lock_vm` now stores a list of VMs it has successfully locked. When done we go through the VMs that we locked and unlock them: this works both when everything succeeds (full list as before), everything fails (empty list), or partial success (we only unlock the VMs that we have previously successfully locked).